_posts/2021-03-23-keccak-optimizer-bug.md

```diff
@@ -0,0 +1,119 @@
+
+```solidity
+contract C {
+  function bug() public returns (uint a, uint b) {
+    assembly {
+      mstore(0, 0)
+      // The optimizer computes the value at compile time:
+      // 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563
+      a := keccak256(0, 32)
+      // The optimizer incorrectly uses the cached value
+      // 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563
+      // instead of 
+      // 0xe2b9f9f9430b05bfa9a3abd3bac9a181434d23a707ef1cde8bd25d30203538d8
+      b := keccak256(0, 23)
+    }
+  }
+}
+```
+
+In the above example, both the Keccak-256 hashes can be computed at compile time. Here are two other
+cases where the hashes cannot be computed at compile time, yet the optimizer considered them equal
+and thus modified the semantics:
+
+```solidity
+contract C {
+  function bug(string memory s) public returns (bool ret) {
+    assembly {
+      let a := keccak256(s, 32)
+      let b := keccak256(s, 8)
+      // Here `a` and `b` were considered equal,
+      // leading to `ret` being incorrectly set to true.
+      ret := eq(a, b)
+    }
+  }
+}
+```
+
+```solidity
+contract C {
+  function bug() public view returns (bool ret) {
+    assembly {
+      let x := calldataload(0)
+      mstore(0, x)
+      mstore(0x20, x)
+      let a := keccak256(0, 4)
+      // even though the memory location is different,
+      // the 32-byte content is the same.
+      let b := keccak256(0x20, 8)
+      // Here `a` and `b` were considered equal,
+      // leading to `ret` being incorrectly set to true.
+      ret := eq(a, b)
+    }
+  }
+}
+```
+
+Specifically, `keccak256(mpos1, length1)` and `keccak256(mpos2, length2)` in some cases were
+considered equal if `length1` and `length2`, when rounded up to nearest multiple of 32 were the
+same, and when the memory contents at `mpos1` and `mpos2` can be deduced to be equal.
+
+You may be **affected** if you compute multiple Keccak-256 hashes of the same content, but with
+different lengths inside inline assembly, and enabled the optimizer.
+
+### Unaffected Cases
+
+You are unaffected in the following situations:
+
+- If your code uses `keccak256` with a length that is not known at compile time.
+- If your code only uses `keccak256` with a length that is always a multiple of 32. In particular,
+  for storage slot computation, the Solidity compiler always computes Keccak-256 hash over memory
+  regions whose lengths and offsets are multiples of 32.
+- If pairs of `keccak256` hashes are separated by certain instructions such as:
+  - An instruction that breaks the control flow, e.g., `jumpi`, `jump` etc. In high level code, this
+     would be function calls, if statements, loops, etc.
+  - An instruction that writes to memory other than `mstore`, e.g., `call`, `returndatacopy`, etc.
+
+### Severity
+
+Since most usages of `keccak256` fall under one of the unaffected cases mentioned in the previous
+section, we assigned the bug a "medium" severity level.
+
```

Comment:
Especially the fact that this can only affect sequences of opcodes without jumps and calls, the likelihood of the bug being present in existing code is very low. This is also validated by the fact that the bug has been present in the optimizer for so long without being detected.

Reply:
Added this.